src/hotspot/os/linux/os_linux.cpp

```diff
@@ -6116,6 +6116,26 @@ bool VM_Crac::is_claimed_fd(int fd) {
   return false;
 }
 
+static void wakeup_threads_in_timedwait() {
+  SafeThreadsListPtr listPtr(Thread::current(), true);
+  ThreadsList *list = listPtr.list();
+  assert(list != NULL, "Thread list is null");
+  for (uint i = 0; i < list->length(); ++i) {
+    JavaThread* t = list->thread_at(i);
+    assert(t != NULL, "Thread is null");
+    ThreadState state = t->osthread()->get_state();
+    // ThreadState::SLEEPING is not used
+    if (state == ThreadState::CONDVAR_WAIT || state == ThreadState::OBJECT_WAIT) {
+      // We want to cause a wakeup but not interrupted exception
+      t->interrupt();
+      t->osthread()->set_interrupted(false);
+    }
```

Comment:
I think the problem of the native implementation of the timed wait is being fixed on the wrong level of abstraction. The interrupt() is much higher leve operation than the pthread_cond_signal, that we are actually targeting. And the latter would be also much cheaper. Could you replace the code with signaling every cond?

Reply:
We are not triggering the Java InterruptedException; when you look into the implementation of `JavaThread::interrupt` you'll see that unparking from those 3 types of conditions is all it's actually doing, setting and immediately unsetting the interrupted flag is a noop on posix systems.
So I don't think this can be any cheaper. We are operating on a bit higher level only to reuse code that we would otherwise inline in here.

Reply:
t->interrupt() triggers unparks of two ParkEvents and of a Parker. They do some bookkeeping in addition to just pthread_cond_timedwait(). But OK, I see we need to actually unpark the thread for park() callers to recalculate the abs time. Anyway, Thread::interrupt looks like a hack. AFAICS unpark()s can be called without Java thread state check, which also will be more straightforward.

Reply:
The thread state check might be more of an optimization, even though unparking a non-parked thread is a noop, it still means some work that's unnecessary.

Reply:
https://github.com/openjdk/crac/blob/crac/src/hotspot/share/runtime/osThread.hpp#L39:
```
// The thread states represented by the ThreadState values are platform-specific
// and are likely to be only approximate, because most OSes don't give you access
// to precise thread state information.

// Note: the ThreadState is legacy code and is not correctly implemented.
// Uses of ThreadState need to be replaced by the state in the JavaThread.
```

Relying on the ThreadState creates a risk of missing unparking, which does not pay off, since the operation is perfromed once on the restore. It's possible that ThreadState will become more broken over time as the rest of Hotspot implementation evolve.